### PR TITLE
Simplicial complex test cases co-authored

### PR DESCRIPTION
**Pull request by maneelusf**

### CODE REVIEW

File: test/nn/simplicial/test_dist2cycle.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+"""Unit tests for Dist2Layer Model."""
+import numpy as np
+import numpy.linalg as npla
+import torch
+from toponetx.classes import SimplicialComplex as sc
+
+from topomodelx.nn.simplicial.dist2cycle import Dist2Cycle
+
+
+class TestDist2Layer:
```
Comment (ninamiolane) [inline test/nn/simplicial/test_dist2cycle.py:10]: Typo: Dist2Cycle, not Dist2Layer?

File: test/nn/simplicial/test_hsn.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+"""Unit tests for HSN Model."""
+import numpy as np
+import numpy.linalg as npla
+import torch
+from toponetx.classes import SimplicialComplex as sc
+
+from topomodelx.nn.simplicial.hsn import HSN
+
+
+class TestDist2Layer:
```
Comment (ninamiolane) [inline test/nn/simplicial/test_hsn.py:10]: Rename: TestHSN

File: test/nn/simplicial/test_sca.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import itertools
+import random
+
+import torch
+from toponetx.classes import SimplicialComplex as sc
+
+from topomodelx.nn.simplicial.sca import AMPSSCA
+
+
+class TestDist2Layer:
```
Comment (ninamiolane) [inline test/nn/simplicial/test_sca.py:11]: Rename

**Review by ninamiolane — CHANGES_REQUESTED** [review]
Impressive work, Thanks! 

Before I look at the code further, please check the names of the Test* Python classes: some of them have the wrong names (I've indicated at some places which ones). They should match the name of the Python class being tested.

THanks!

**Comment by maneelusf** [comment]
@ninamiolane , thank you! I have resolved them. 
On another note,
sccconv(simplicial 2-complex CNN) i.e. SCConv has not been implemented(see SS below). Also, there are a couple of models(ScoNe/HSN) for which the loss is not reducing in the notebook implementation. Not sure if that is the right implementation or the intended behavior.

Screenshot 2023-09-07 at 4 54 46 PM

File: test/nn/simplicial/test_dist2cycle.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+"""Unit tests for Dist2Cycke Model."""
+import numpy as np
+import numpy.linalg as npla
```
Comment (ninamiolane) [inline test/nn/simplicial/test_dist2cycle.py:3]: Remove this line and use np.linalg. everytime you need it.

Best practices are to avoid renaming module elements, because it hides where the elements, functions, etc are coming from when we are reading the code.

File: test/nn/simplicial/test_dist2cycle.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import torch
+from toponetx.classes import SimplicialComplex as sc
```
Comment (ninamiolane) [inline test/nn/simplicial/test_dist2cycle.py:5]: Ditto: Remove `as sc`
(and adapt the code accordingly)

File: test/nn/simplicial/test_dist2cycle.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    def test_forward(self):
+        """Test the forward method of Dist2Cycle."""
+        edge_set = [[1, 2], [1, 3], [2, 5], [3, 5]]
+        face_set = [[2, 3, 4], [2, 4, 5]]
+
+        torch.manual_seed(42)
+        ex2_sc = sc(edge_set + face_set)
+        ld = ex2_sc.down_laplacian_matrix(rank=1).todense()
+        A = ex2_sc.adjacency_matrix(rank=1).todense()
```
Comment (ninamiolane) [inline test/nn/simplicial/test_dist2cycle.py:21]: TopoModelX does not use the convention `A` to refer to the adjacency matrix.
For example:
https://github.com/pyt-team/TopoModelX/blob/486adf2299eb37aa22ec1fdf54351458074da253/topomodelx/nn/simplicial/hsn_layer.py#L114

Here, you should write `adjacency_1`.

Note: In Python, uppercase letters are reserved for class names (observe how they get colored in orange: orange means Python class). Thus, we should not use uppercase letters for variable names.

File: test/nn/simplicial/test_dist2cycle.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    """Unit tests for the Dist2Cycle model class."""
+
+    def test_forward(self):
+        """Test the forward method of Dist2Cycle."""
+        edge_set = [[1, 2], [1, 3], [2, 5], [3, 5]]
+        face_set = [[2, 3, 4], [2, 4, 5]]
+
+        torch.manual_seed(42)
+        ex2_sc = sc(edge_set + face_set)
+        ld = ex2_sc.down_laplacian_matrix(rank=1).todense()
```
Comment (ninamiolane) [inline test/nn/simplicial/test_dist2cycle.py:20]: `ld` is not the convention used in TopoModelX to refer to the laplacian.
See for example:
https://github.com/pyt-team/TopoModelX/blob/486adf2299eb37aa22ec1fdf54351458074da253/topomodelx/nn/simplicial/sccnn_layer.py#L266

In your case here, use instead: `laplacian_down_1.

Note: In Python, it is best practice to avoid short variable names. Because: (1) they are not informative for users / readers of the code and make everything confusing. (2) "Find and replace" on short variable names would change a lot of things in the code. Not convenient.

File: test/nn/simplicial/test_dist2cycle.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+class TestDist2Cycle:
+    """Unit tests for the Dist2Cycle model class."""
+
+    def test_forward(self):
+        """Test the forward method of Dist2Cycle."""
+        edge_set = [[1, 2], [1, 3], [2, 5], [3, 5]]
+        face_set = [[2, 3, 4], [2, 4, 5]]
+
+        torch.manual_seed(42)
+        ex2_sc = sc(edge_set + face_set)
```
Comment (ninamiolane) [inline test/nn/simplicial/test_dist2cycle.py:19]: ex2_sc is not an informative name. We need names that help the reader/user read the code as if it was a text. Use:

```suggestion
        simplicial_complex = SimplicialComplex(edge_set + face_set)
```

File: test/nn/simplicial/test_dist2cycle.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        ld = ex2_sc.down_laplacian_matrix(rank=1).todense()
+        A = ex2_sc.adjacency_matrix(rank=1).todense()
+        L_tilde_pinv = npla.pinv(ld + np.eye(ld.shape[0]))  # test inverse
```
Comment (ninamiolane) [inline test/nn/simplicial/test_dist2cycle.py:22]: Ditto: find better names by: (1) reading my comments above naming above, (2) looking in the codebase of TopoModelX to observe which conventions are chosen, and respect the same ones.

File: test/nn/simplicial/test_hsn.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
+"""Unit tests for HSN Model."""
+import numpy as np
+import numpy.linalg as npla
+import torch
+from toponetx.classes import SimplicialComplex as sc
```
Comment (ninamiolane) [inline test/nn/simplicial/test_hsn.py:5]:
```suggestion
from toponetx.classes import SimplicialComplex
```

File: test/nn/simplicial/test_hsn.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
+"""Unit tests for HSN Model."""
+import numpy as np
+import numpy.linalg as npla
```
Comment (ninamiolane) [inline test/nn/simplicial/test_hsn.py:3]:
```suggestion
```

File: test/nn/simplicial/test_hsn.py (hunk 2 above)
Comment (ninamiolane) [reply test/nn/simplicial/test_hsn.py:10]: This is not resolved. The name still shows TestDist2Layer instead of HSN.

**Review by ninamiolane — CHANGES_REQUESTED** [review]
- [ ] Some Python class names have not been changed. See test_hsn.py for example. I have unresolved that conversation. Could you check again and make sure that the class names are good?
- [ ] The names you chose for variables (e.g. `ld` or `A`) (1) do not respect Python code standards, (2) do not respect the conventions chosen in TopoModelX. I have put some comments in the code review that explain this. Please change the variable names in every file.
- [ ] Remove the files: test_sconebis.py and scone_bis.py: they were deleted in another PR.

Thanks for this very substantial work. This is an amazing contribution to the software.

**Comment by ninamiolane** [comment]
@maneelusf Thanks for your comments about SCConv not implemented and models(ScoNe/HSN) for which the loss is not reducing in the notebook implementation. I have made notes of these, we will have a look.

**Comment by mhajij** [comment]
@maneelusf can you please take a look at the comments above and try to address them asap?

File: test/nn/simplicial/test_san.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        in_channels = x.shape[-1]
+        hidden_channels = 16
+        out_channels = 2
+        model = SAN(
+            in_channels=in_channels,
+            hidden_channels=hidden_channels,
+            out_channels=out_channels,
+            n_layers=1,
+        )
+        Ldown = torch.from_numpy(
```
Comment (ninamiolane) [inline test/nn/simplicial/test_san.py:48]:
```suggestion
        laplacian_down_1 = torch.from_numpy(
```

File: test/nn/simplicial/test_san.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+            simplicial_complex.down_laplacian_matrix(rank=1).todense()
+        ).to_sparse()
+        Lup = torch.from_numpy(
```
Comment (ninamiolane) [inline test/nn/simplicial/test_san.py:51]:
```suggestion
        laplacian_up_1 = torch.from_numpy(
```

File: test/nn/simplicial/test_san.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+            simplicial_complex.up_laplacian_matrix(rank=1).todense()
+        ).to_sparse()
+
+        assert torch.any(
+            torch.isclose(
+                model(x, Lup, Ldown)[0], torch.tensor([0.7727, 0.2389]), rtol=1e-02
```
Comment (ninamiolane) [inline test/nn/simplicial/test_san.py:57]:
```suggestion
                model(x, laplacian_up_1, laplacian_down_1)[0], torch.tensor([0.7727, 0.2389]), rtol=1e-02
```

File: test/nn/simplicial/test_sca.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        random.shuffle(all_combinations)
+        selected_combinations = all_combinations[:faces]
+        simplicial_complex = SimplicialComplex()
+        for simplex in selected_combinations:
+            simplicial_complex.add_simplex(simplex)
+
+        x_0 = torch.randn(15, 2)
+        x_1 = torch.randn(35, 2)
+        x_2 = torch.randn(14, 2)
+        down_lap1 = simplicial_complex.down_laplacian_matrix(rank=1)
```
Comment (ninamiolane) [inline test/nn/simplicial/test_sca.py:37]: Please use conventions for names on laplacians.

`laplacian_down_1`.

Note that you can easily and quickly replace all of these with code editors like visual code.

**Review by ninamiolane — APPROVED** [review]
Thank you very much for this huge work. There were some (very minor) name conventions problems, I've fixed a few, and we will solve the rest of in a future PR: let's merge this! 

Well done 🌟
